macOS/UnitTests/NetworkProtection/VPNUpsellPopoverViewModelTests.swift

```diff
@@ -0,0 +1,285 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import XCTest
+import Combine
+import Common
+import VPN
+import BrowserServicesKit
+import SubscriptionTestingUtilities
+import Subscription
+@testable import DuckDuckGo_Privacy_Browser
+
+@MainActor
+final class VPNUpsellPopoverViewModelTests: XCTestCase {
+    var sut: VPNUpsellPopoverViewModel!
+    var mockSubscriptionManager: SubscriptionAuthV1toV2BridgeMock!
+    var mockFeatureFlagger: MockFeatureFlagger!
+    var mockDefaultBrowserProvider: MockDefaultBrowserProvider!
+    var mockPersistor: MockVPNUpsellUserDefaultsPersistor!
+    var vpnUpsellVisibilityManager: VPNUpsellVisibilityManager!
+    var lastReceivedURL: URL?
+    var cancellables: Set<AnyCancellable> = []
+
+    override func setUp() {
+        super.setUp()
+        mockSubscriptionManager = SubscriptionAuthV1toV2BridgeMock()
+        mockFeatureFlagger = MockFeatureFlagger()
+        mockDefaultBrowserProvider = MockDefaultBrowserProvider()
+        mockPersistor = MockVPNUpsellUserDefaultsPersistor()
+
+        mockFeatureFlagger.enabledFeatureFlags = [.vpnToolbarUpsell]
+
+        vpnUpsellVisibilityManager = VPNUpsellVisibilityManager(
+            isFirstLaunch: false,
+            isNewUser: true,
+            subscriptionManager: mockSubscriptionManager,
+            defaultBrowserProvider: mockDefaultBrowserProvider,
+            contextualOnboardingPublisher: Just(true).eraseToAnyPublisher(),
+            featureFlagger: mockFeatureFlagger,
+            persistor: mockPersistor,
+            timerDuration: 0.01,
+            autoDismissDays: 7
+        )
+        vpnUpsellVisibilityManager.setup(isFirstLaunch: false)
+
+        sut = VPNUpsellPopoverViewModel(
+            subscriptionManager: mockSubscriptionManager,
+            featureFlagger: mockFeatureFlagger,
+            vpnUpsellVisibilityManager: vpnUpsellVisibilityManager,
+            urlOpener: { url in
+                self.lastReceivedURL = url
+            },
+            onDismiss: {}
+        )
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        sut = nil
+        vpnUpsellVisibilityManager = nil
+        mockSubscriptionManager = nil
+        mockFeatureFlagger = nil
+        mockDefaultBrowserProvider = nil
+        lastReceivedURL = nil
+        mockPersistor = nil
+        cancellables.removeAll()
+    }
+
+    func testWhenPopoverIsDismissed_ThenDismissedFlagIsSet() throws {
+            // Given
+            XCTAssertEqual(vpnUpsellVisibilityManager.state, .visible)
+            XCTAssertFalse(mockPersistor.vpnUpsellDismissed)
+
+            // When
+            sut.dismiss()
+
+            // Then
+            XCTAssertTrue(mockPersistor.vpnUpsellDismissed)
+        XCTAssertEqual(vpnUpsellVisibilityManager.state, .dismissed)
+    }
+
+    func testWhenPrimaryCTAIsClicked_SubscriptionLandingPageIsOpened_AndOriginIsSet() throws {
```

Comment:
@aataraxiaa I decided to move the deision logic for the variants into the VM (eg. “is the user eligible for free trial?”, or “is paid AI enabled?”, and added a bunch of tests here.